### ISSUE TITLE
Column Wrap

### ISSUE THREAD

**Issue by donnnyrewq** [techprimate/TPPDF#113]
Is it possible to have content wrap to the next column instead of the next page? Or to get the height of each added object (text, table, etc) so I can manually wrap object to the next column in a section?

**Comment by techprimate-phil** [comment]
Hi, this should be possible with Multi-Column Sections. Please see the [README](https://github.com/techprimate/TPPDF#multi-column-sections)

I will close this for now, if this does not fit your need, please feel free to reopen

**Comment by pj4533** [comment]
Could you provide more detail here?  I am trying to do something similar.   I have a large attributed string, and I want a 2 column layout for my PDF.   If I create a multi-column section as mentioned in the README, it requires me to specify the column for the attributed string, rather than 'wrapping' to the second column.

**Comment by techprimate-phil** [comment]
To clarify:

- You want two columns per page
- You want to add content as usual
- When the column one is full it continues in column 2
- When both columns are full it should add a page break and continue in column one again

Is this correct? If yes, #9 left off with this feature.

**Comment by pj4533** [comment]
@techprimate-phil Correct, yes.  I guess that would be described as “floating” or “spanning” the content between columns?   But you described my use-case exactly, I have a large amount of text (potentially many pages), and I want one column to fill up, the content to go to the next column, then at the end of the second column, insert a page break.

**Comment by techprimate-phil** [comment]
This exact functionality is currently not implemented, but should definitely be part of this framework.

Currently multi-sections are used to split a page vertically, and adding content to each column individually.

I guess adding this feature to the `MultiColumnSection` might not be appropriate, because we would only define the amount of columns and keep adding the content to the document, not to a specific column.

I propose we create something like a `MultiColumnWrapperSection` with a `columns: Int` property. It should only require us to do calculations and creating a user-friendly interface for the library users

**Comment by techprimate-phil** [comment]
I just took a shot at implementing this feature, and looks like it works.

Please checkout the feature branch `feature/ColumnWrapSection`.
The following example should work:

```swift
document.addText(text: "Text 1")
document.addText(text: "Text 2")
document.addText(text: "Text 3")
document.addText(text: "Text 4")
document.addText(text: "Text 5")
document.enable(columns: 4);
for i in 0..<200 {
    document.addText(text: "\(i)-\(i)-\(i)-\(i)")
}
document.disableColumns();
```

If this solves your issue, would be great to hear some feedback.

**Comment by pj4533** [comment]
In general, looks great.   Few things:

- I really like the flexibility of the percentage column widths.   Specifically to make a "margin" column in the middle.   For example, I have my text columns at [0.48,0.04,0.48],  as I found that gives me a nice middle margin between text columns.

- It appears when the columns wrap, the rightmost column doesn't take into account the previous height of the text.   For example (using your code):

<img width="1187" alt="image" src="https://user-images.githubusercontent.com/110283/58491707-807a5600-813d-11e9-92df-d320cbff142b.png">

But like I said, in general, this is exactly my use case.   Thanks so much for the quick turnaround!!!

**Comment by pj4533** [comment]
To be clear, I would expect the '105' line to be at the same level at the '70' line.

**Comment by techprimate-phil** [comment]
You are welcome, glad you like it, it really is overdue.

the alignment error is definitely unintentional. I‘ll take a look and fix it.

I thought about the dynamic column widths too, needs more calculations but should be fine.

I think we need to add an extra spacing (array) parameter instead of „abusing“ empty columns, as it could be misleading.

But I am not entirely sure about best syntax, as the lengths of the column-widths array and the spacing-array will differ by one (e.g. 4 columns, 3 spacings) and I do try to keep the framework as simple as possible.

**Comment by techprimate-phil** [comment]
I just pushed an update fixing the layout issue of the last column and adding relative column widths and spacing.

Example:

```swift
for i in 0..<10 {
    document.addText(text: "Text \(i)")
}
document.enable(columns: 4, widths: [0.2, 0.3, 0.4, 0.1], spacings: [10, 50, 20]);
for i in 0..<200 {
    document.addText(text: "\(i)-A-B-C-D-E-F-G-H-I-J-K-L-M-N-O-P-Q-R")
}
document.disableColumns();
for i in 0..<10 {
    document.addText(text: "Text \(i)")
}
```

**Comment by pj4533** [comment]
I tested it in my project and it works great!  Thanks again.

The last thing I'd mention, and this might be a bigger task is.....images.   Another portion of my PDF use-case is to output UIViews to PDF in a double column.  I do this by rendering to image and doing:

```swift
section.columns[column].addImage(.center, image: imageElement)
```

And manually altering the column myself based on height (similar to what I previously did for text).  In a PERFECT implementation, it would actually wrap the image itself, so if the image was taller than a full page (def possible in my app), it would continue on the next column (or next page depending).   However, currently documents don't even do that (they just cut off the image), so I handle that as best I can.   The key feature would be allowing me to get rid of the logic for managing columns manually and just doing an 'addImage'   when in column mode.   

tl;dr   Text works great, any chance of images working?

**Comment by techprimate-phil** [comment]
One issue I faced yesterday is the broken indentations inside the columns. I'll take a look if I can combine columns with custom indentations, as it would resolve a use case of myself. Currently the column calculations adds `PDFIndentationsObjects` into the render list, therefore removing any custom indentation logic.

About image wrapping, can you provide me a code example or at least a reference image? 
It would help me with understanding the exact use case.

**Comment by pj4533** [comment]
Yes, I might run into that indentation issue if I decide to go with a slightly different format for my output.  As of now, I don't have that particular issue.

As the the image wrapping, it might be very specific to my case, so likely a low priority (but never hurts to mention it!).   Anyway, here is an example page from my app.  The left column is a UIView that I rendered to a UIImage.  The desired width for my UIView is around 400-420 or so.  However, at that width, several times the content is too tall, as is the case here:

[desired_width.pdf](https://github.com/techprimate/TPPDF/files/3236976/desired_width.pdf)

What I'd like to have happen is the image 'wraps' to the top of the following column.   For now, I have hacked it by just making the width of my UIView wider, so my content is shorter.  But that makes it less readable:

[wider_to_fit.pdf](https://github.com/techprimate/TPPDF/files/3236997/wider_to_fit.pdf)

Plus, there _could_ be content that is taller, even at that width, so its not a good solution for me.

**Comment by techprimate-phil** [comment]
Might take a few days, I'll come back to this issue.

Alright, looks like a very specific problem. I do not think that image wrapping with this behaviour should be part of this framework (to keep its usage simplicity), especially we would need a very advanced logic on how to crop/slice the images, which seems out of scope for now.

All I can recommend is either calculating the correct image size and preparing the images before creating the PDF, or generating the whole content using TPPDF instead of UIView rendered images.

**Comment by pj4533** [comment]
Cool, sounds good.  I agree about the images, would be complicated fast.  Thanks again!

**Comment by techprimate-phil** [comment]
Well implementing went quicker as expected, checkout the example in the feature branch.

I'll prepare an release as mentioned in the upcoming days, maybe creating some unit tests for it first

**Comment by techprimate-phil** [comment]
As I am merging all changes from the past few days, I realised that the column wrap section, messes up the header and footer calculations. Therefore not release ready yet, but it is available on the `develop` branch now

**Comment by pj4533** [comment]
👍

**Comment by techprimate-phil** [comment]
alright should be fixed, preparing a release now

**Comment by pj4533** [comment]
Awesome.   I am aaaaaaaalmost ready to ship v1 of my app so its perfect timing.

**Comment by techprimate-phil** [comment]
Implemented in release 1.5.0

**Comment by pj4533** [comment]
Woohoo!  Updated, working great.
